superset/security/manager.py

```diff
@@ -2063,6 +2064,27 @@ def can_access_based_on_dashboard(datasource: "BaseDatasource") -> bool:
         )
 
         exists = db.session.query(query.exists()).scalar()
+
+        # check for datasets that are only used by filters
+        if not exists:
+            dashboards_json = (
+                db.session.query(Dashboard.json_metadata)
+                .filter(Dashboard.id.in_(dashboard_ids))
+                .all()
+            )
+            for json_ in dashboards_json:
+                try:
+                    json_metadata = json.loads(json_.json_metadata)
+                    for filter_ in json_metadata.get("native_filter_configuration", []):
+                        filter_dataset_ids = [
+                            target.get("datasetId")
+                            for target in filter_.get("targets", [])
+                        ]
+                        if datasource.id in filter_dataset_ids:
+                            exists = True
```

Comment:
```suggestion
                            return True
```

We should short circuit when possible.

Reply:
Makes perfect sense! Made this change.